enable levels to override variables of inheriting levels

vojtapolasek opened the pull request:
#### Description:

- modify build system
- while getting list of controls of a certain level, make sure that only the top-most level variable definition is used


#### Rationale:

- this modification allows for example CIS level 2 profile to have a different variable than CIS level 1 profile

pep8speaks commented:
Hello @vojtapolasek! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`ssg/controls.py`](https://github.com/ComplianceAsCode/content/blob/aac161bf74fd569c1a5d4de397e918f8ea8b653b/ssg/controls.py):

> [Line 208:13](https://github.com/ComplianceAsCode/content/blob/aac161bf74fd569c1a5d4de397e918f8ea8b653b/ssg/controls.py#L208): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'

alexhaydock commented:
I don't really have the deep level of CaC understanding to work out exactly how this applies but just to check - will variables set in tailoring files still take precedence even when this change is merged?

cipherboy commented:
@alexhaydock Tailoring files are applied at the XCCDF level. You can think of CaC as one big build system to generate XCCDFs (and other artifacts). As long as @vojtapolasek doesn't rename the variables (unlikely), there's no impact to tailoring files. This just changes how the Controls format (a CaC-ism used to generate Profiles) selects the default value of a variable at a given level.

yuumasato commented on ssg/controls.py:
```diff
@@ -202,9 +202,16 @@ def get_all_controls_of_level(self, policy_id, level_id):
-            if len(level_ids.intersection(c.levels)) > 0:
-                eligible_controls.append(c)
+        defined_variables = []
+        # we will go level by level, from top to bottom
+        # this is done to enable overriding of variables by higher levels
+        for lv in level_ids:
```
`level_ids` is a `set`, so we cannot assume any particular order over the IDs.

Additionally, we cannot assume that the levels returned by `get_level_with_ancestors` is ordered, it also returns the levels in a `set`.

yuumasato reviewed (changes requested):
I gave it quick try with ANSSI profiles and it doesn't work.
I took a random variable from the profile, included it in a control from each level and assigned each one a different  selector. The variables came assigned with incorrect values, and each time I built the content the assignments varied.

Maybe we need to convert those sets to a list? I don't know what else would be impacted.

It would be good to have tests for this as well.

vojtapolasek commented:
Hello, I updated the PR. I replaced set with ordereddict. I also added a new functionality which can explicitly enable level overriding. I am not sure if the functionality is in the right part of the build system.
@yuumasato  @matejak  @cipherboy  what do you think?

vojtapolasek commented:
I have updated the PR. I made overriding of variables hardcoded and added some tests. @yuumasato  @jan-cerny  what do you think now?

vojtapolasek commented:
I think I fixed it.

matejak commented on ssg/controls.py:
```diff
@@ -198,13 +201,32 @@ def _get_policy(self, policy_id):
+        # this is done to enable overriding of variables by higher levels
+        for lv in levels.keys():
+            for c in all_policy_controls:
+                if lv.id in c.levels:
```
This `if` can be simplified by `if lv.id not in c.levels: continue` - one level of indentation less.

The following code is not complex, but it needs extraction into meaningfully named private functions. Right now, it is very difficult to guess what are we doing why.

yuumasato reviewed (approved):
LGTM, thank you for the enhancement, :)

matejak commented:
I like the test, and I believe that the code works, but I would like to see some code readability improvements. As the original author won't be available for some time, I can tackle that next week.

matejak commented:
I have made following adjustments:

- Renamed get_level_with_ancestors to get_level_with_ancestors_sequence, and made it return a list instead - a dictionary result is quite confusing.
- Removed some optimization in the variable deletion loops.
- Extracted functionality to a _get_control_without_variables static method.
- Defined variable removal steps using set operations.

yuumasato reviewed (approved):
@matejak Thank you for the improvements.
Looks great now.
